xdsl/transforms/memref_to_dsd.py

```diff
@@ -0,0 +1,203 @@
+from dataclasses import dataclass
+from typing import cast
+
+from xdsl.context import MLContext
+from xdsl.dialects import arith, builtin, csl, memref
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    Float16Type,
+    Float32Type,
+    IntAttr,
+    IntegerAttr,
+    IntegerType,
+    MemRefType,
+    ModuleOp,
+    Signedness,
+    StridedLayoutAttr,
+)
+from xdsl.ir import Attribute, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.hints import isa
+
+
+class LowerAllocOpPass(RewritePattern):
```

Comment:
I'm a bit concerned this will put the array in the function body (and whether that is OK wrt exporting)

Reply:
In the interest of writing simple transforms, this translates everything in-place without analysing context. MLIR has an extra pass in bufferization for `--buffer-hoisting`. Not sure if we can use it as-is, but it might be simpler to have this logic somewhere separate.